Fix font preview text color on light background

timothyqiu opened the pull request:
Pick white/black text color based on the luminance of background.

| Before | After |
|----------|--------|
| ![before-1](https://user-images.githubusercontent.com/372476/146317063-2e3e2dcf-26b1-4a00-9514-c61e6f38dbcc.png) | ![after-1](https://user-images.githubusercontent.com/372476/146317090-c6dffc34-e356-491f-bef8-ff4de1314501.png) |
| ![before-2](https://user-images.githubusercontent.com/372476/146317110-b74f4c50-bcf1-4271-ab03-5947e85c9649.png) | ![after-2](https://user-images.githubusercontent.com/372476/146317154-5ee35acc-c05a-42a9-a07b-553c7886c8d4.png) |
| ![before-3](https://user-images.githubusercontent.com/372476/146317194-b2428878-3d95-477d-ab68-8aa6f2916c30.png) | ![after-3](https://user-images.githubusercontent.com/372476/146317204-d0f50b70-f8f1-4335-a2e8-da75b482c027.png) |
| ![before-4](https://user-images.githubusercontent.com/372476/146317227-1b29bd20-bb22-4a60-bcfc-f631ef7e13b0.png) | ![after-4](https://user-images.githubusercontent.com/372476/146317236-eb293afe-0872-4855-a9ab-382b02229a67.png) |

timothyqiu committed:
Fix font preview text color on light background

akien-mga commented on editor/plugins/editor_preview_plugins.cpp:
```diff
@@ -851,7 +851,11 @@ Ref<Texture2D> EditorFontPreviewPlugin::generate_from_path(const String &p_path,
 
-	font->draw_string(canvas_item, pos, sample, HORIZONTAL_ALIGNMENT_LEFT, -1.f, 50, Color(1, 1, 1));
+	const Color c = GLOBAL_GET("rendering/environment/defaults/default_clear_color");
+	const float luminance = 0.2126 * c.r + 0.7152 * c.g + 0.0722 * c.b;
```
I'm surprised that we wouldn't have something like this in `Color` already.

Does that mean that other parts of the adaptive theme coloring also hardcode similar checks, or use a different metric? CC @pycbouh @Calinou

pycbouh replied:
Indeed, a different approach is used, the color components are averaged and that value is checked:

```cpp
bool EditorSettings::is_dark_theme() {
	int AUTO_COLOR = 0;
	int LIGHT_COLOR = 2;
	Color base_color = get("interface/theme/base_color");
	int icon_font_color_setting = get("interface/theme/icon_and_font_color");
	return (icon_font_color_setting == AUTO_COLOR && ((base_color.r + base_color.g + base_color.b) / 3.0) < 0.5) || icon_font_color_setting == LIGHT_COLOR;
}
```

I'm not strong with the color theory, but it makes sense to unify this somehow. What would the `Color` method be? `is_dark()/is_light()` that returns true depending on the luminance factor compared to `0.5`?

Zireael07 replied:
Some people use rgb, some use luminance, personally I just hardcoded things, but this is definitely something worth unifying and having built-in (AND exposed to scripting!)

timothyqiu commented:
Added `Color.get_luminance()` instead of `is_dark()`/`is_light()`:

* Avoids to pick a name 😛 as "luminance" is a recognized property of color
* Avoids to decide whether 0.5 is dark or light.

Also changed `EditorSettings::is_dark_theme()` to use `get_luminance()`.

pycbouh commented:
> Avoids to decide whether 0.5 is dark or light. 

Well, depends on how you look at it 😛 Now each caller needs to decide on how to interpret the value. But sure, it works. Maybe worth a note in the docs to explain why you'd use it and that 0.5 is a good cutoff for deciding about whether the color is dark or light.

YuriSizov reviewed (approved):
Changes make sense 👍

akien-mga commented:
Thanks!
